File: frontend/nfe/lib/app/data/model/nfe_cupom_fiscal_referenciado_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:intl/intl.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeCupomFiscalReferenciadoModel {
	int? id;
	int? idNfeCabecalho;
	String? modeloDocumentoFiscal;
	int? numeroOrdemEcf;
	int? coo;
	DateTime? dataEmissaoCupom;
	int? numeroCaixa;
	String? numeroSerieEcf;

	NfeCupomFiscalReferenciadoModel({
		this.id,
		this.idNfeCabecalho,
		this.modeloDocumentoFiscal,
		this.numeroOrdemEcf,
		this.coo,
		this.dataEmissaoCupom,
		this.numeroCaixa,
		this.numeroSerieEcf,
	});

	static List<String> dbColumns = <String>[
		'id',
		'modelo_documento_fiscal',
		'numero_ordem_ecf',
		'coo',
		'data_emissao_cupom',
		'numero_caixa',
		'numero_serie_ecf',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Modelo Documento Fiscal',
		'Numero Ordem Ecf',
		'Coo',
		'Data Emissao Cupom',
		'Numero Caixa',
		'Numero Serie Ecf',
	];

	NfeCupomFiscalReferenciadoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCabecalho = jsonData['idNfeCabecalho'];
		modeloDocumentoFiscal = NfeCupomFiscalReferenciadoDomain.getModeloDocumentoFiscal(jsonData['modeloDocumentoFiscal']);
		numeroOrdemEcf = jsonData['numeroOrdemEcf'];
		coo = jsonData['coo'];
		dataEmissaoCupom = jsonData['dataEmissaoCupom'] != null ? DateTime.tryParse(jsonData['dataEmissaoCupom']) : null;
		numeroCaixa = jsonData['numeroCaixa'];
		numeroSerieEcf = jsonData['numeroSerieEcf'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCabecalho'] = idNfeCabecalho != 0 ? idNfeCabecalho : null;
		jsonData['modeloDocumentoFiscal'] = NfeCupomFiscalReferenciadoDomain.setModeloDocumentoFiscal(modeloDocumentoFiscal);
		jsonData['numeroOrdemEcf'] = numeroOrdemEcf;
		jsonData['coo'] = coo;
		jsonData['dataEmissaoCupom'] = dataEmissaoCupom != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissaoCupom!) : null;
		jsonData['numeroCaixa'] = numeroCaixa;
		jsonData['numeroSerieEcf'] = numeroSerieEcf;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCabecalho = plutoRow.cells['idNfeCabecalho']?.value;
		modeloDocumentoFiscal = plutoRow.cells['modeloDocumentoFiscal']?.value != '' ? plutoRow.cells['modeloDocumentoFiscal']?.value : 'AAA';
		numeroOrdemEcf = plutoRow.cells['numeroOrdemEcf']?.value;
		coo = plutoRow.cells['coo']?.value;
		dataEmissaoCupom = Util.stringToDate(plutoRow.cells['dataEmissaoCupom']?.value);
		numeroCaixa = plutoRow.cells['numeroCaixa']?.value;
		numeroSerieEcf = plutoRow.cells['numeroSerieEcf']?.value;
	}	

	NfeCupomFiscalReferenciadoModel clone() {
		return NfeCupomFiscalReferenciadoModel(
			id: id,
			idNfeCabecalho: idNfeCabecalho,
			modeloDocumentoFiscal: modeloDocumentoFiscal,
			numeroOrdemEcf: numeroOrdemEcf,
			coo: coo,
			dataEmissaoCupom: dataEmissaoCupom,
			numeroCaixa: numeroCaixa,
			numeroSerieEcf: numeroSerieEcf,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_cupom_fiscal_referenciado_model.py
from src import db


class NfeCupomFiscalReferenciadoModel(db.Model):
    __tablename__ = 'nfe_cupom_fiscal_referenciado'

    id = db.Column(db.Integer, primary_key=True)
    modelo_documento_fiscal = db.Column(db.String(2))
    numero_ordem_ecf = db.Column(db.Integer)
    coo = db.Column(db.Integer)
    data_emissao_cupom = db.Column(db.DateTime)
    numero_caixa = db.Column(db.Integer)
    numero_serie_ecf = db.Column(db.String(21))
    id_nfe_cabecalho = db.Column(db.Integer, db.ForeignKey('nfe_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cabecalho = data.get('idNfeCabecalho')
        self.modelo_documento_fiscal = data.get('modeloDocumentoFiscal')
        self.numero_ordem_ecf = data.get('numeroOrdemEcf')
        self.coo = data.get('coo')
        self.data_emissao_cupom = data.get('dataEmissaoCupom')
        self.numero_caixa = data.get('numeroCaixa')
        self.numero_serie_ecf = data.get('numeroSerieEcf')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCabecalho': self.id_nfe_cabecalho,
            'modeloDocumentoFiscal': self.modelo_documento_fiscal,
            'numeroOrdemEcf': self.numero_ordem_ecf,
            'coo': self.coo,
            'dataEmissaoCupom': self.data_emissao_cupom.isoformat(),
            'numeroCaixa': self.numero_caixa,
            'numeroSerieEcf': self.numero_serie_ecf,
        }